src/compiler/lambda.rs

```diff
@@ -0,0 +1,164 @@
+use std::borrow::Borrow;
+use std::rc::Rc;
+
+use num_bigint::ToBigInt;
+
+use crate::compiler::clvm::truthy;
+use crate::compiler::comptypes::{BodyForm, CompileErr, CompilerOpts, PrimaryCodegen};
+use crate::compiler::evaluate::{make_operator1, make_operator2};
+use crate::compiler::frontend::{compile_bodyform, frontend};
+use crate::compiler::sexp::{enlist, SExp};
+use crate::compiler::srcloc::Srcloc;
+
+pub fn compose_constant_function_env(compiler: &PrimaryCodegen) -> Result<Rc<SExp>, CompileErr> {
+    match compiler.env.borrow() {
+        SExp::Cons(_, left, _) => Ok(left.clone()),
+        _ => Ok(Rc::new(SExp::Nil(compiler.env.loc()))),
+    }
+}
+
+fn make_captures(opts: Rc<dyn CompilerOpts>, sexp: Rc<SExp>) -> Result<Rc<BodyForm>, CompileErr> {
+    if let SExp::Cons(l, f, r) = sexp.borrow() {
+        Ok(Rc::new(make_operator2(
+            l,
+            "c".to_string(),
+            make_captures(opts.clone(), f.clone())?,
+            make_captures(opts, r.clone())?,
+        )))
+    } else if !truthy(sexp.clone()) {
+        Ok(Rc::new(BodyForm::Quoted(SExp::Nil(sexp.loc()))))
+    } else {
+        Ok(Rc::new(compile_bodyform(opts, sexp)?))
+    }
+}
+
+fn find_and_compose_captures(
+    opts: Rc<dyn CompilerOpts>,
+    sexp: &SExp,
+) -> Result<(Rc<SExp>, Rc<BodyForm>), CompileErr> {
+    let mut args = Rc::new(sexp.clone());
+    let mut capture_args = Rc::new(SExp::Nil(sexp.loc()));
+    let mut captures = Rc::new(BodyForm::Quoted(SExp::Nil(sexp.loc())));
+    if let SExp::Cons(_, l, r) = sexp {
+        if let SExp::Cons(_, head, rest) = l.borrow() {
+            if let SExp::Atom(_, name) = head.borrow() {
+                if name == b"&" {
+                    args = r.clone();
+                    capture_args = rest.clone();
+                    captures = make_captures(opts, rest.clone())?;
+                }
+            }
+        }
+    }
+
+    Ok((
+        Rc::new(SExp::Cons(sexp.loc(), capture_args, args)),
+        captures,
+    ))
+}
+
+fn make_call(loc: Srcloc, head: &str, args: &[BodyForm]) -> BodyForm {
+    let mut use_vec: Vec<Rc<BodyForm>> = args.iter().cloned().map(Rc::new).collect();
+    use_vec.insert(
+        0,
+        Rc::new(BodyForm::Value(SExp::atom_from_string(loc.clone(), head))),
+    );
+    BodyForm::Call(loc, use_vec)
+}
+
+fn make_operator(loc: Srcloc, op: u8, arg1: Rc<BodyForm>, arg2: Rc<BodyForm>) -> BodyForm {
+    BodyForm::Call(
+        loc.clone(),
+        vec![
+            Rc::new(BodyForm::Value(SExp::Atom(loc, vec![op]))),
+            arg1,
+            arg2,
+        ],
+    )
+}
+
+fn make_cons(loc: Srcloc, arg1: Rc<BodyForm>, arg2: Rc<BodyForm>) -> BodyForm {
+    make_operator(loc, 4, arg1, arg2)
+}
+
+//
+// Lambda
+//
+// (lambda ((= captures) arguments)
```

Comment:
I got the impression the syntax is to use `&` for capture expressions, not `=`. or am I misreading this comment?